File: OrangeJuice.Server.Core/Configuration/AzureAwsOptionsProvider.cs
﻿using System.Linq;
using System.Threading.Tasks;

using Microsoft.WindowsAzure.Storage.Table;

using OrangeJuice.Server.Services;

namespace OrangeJuice.Server.Configuration
{
	public sealed class AzureAwsOptionsProvider : IOptionsProvider<AwsOptions>
	{
		private readonly AzureOptions _azureOptions;
		private readonly IAzureClient _azureClient;
		private readonly IConverter<DynamicTableEntity, AwsOptions> _converter;

		public AzureAwsOptionsProvider(AzureOptions azureOptions, IAzureClient azureClient, IConverter<DynamicTableEntity, AwsOptions> converter)
		{
			_azureOptions = azureOptions;
			_azureClient = azureClient;
			_converter = converter;
		}

		public async Task<AwsOptions[]> GetOptions()
		{
			var content = await _azureClient.GetEntitiesFromTable<DynamicTableEntity>(_azureOptions.AwsOptionsTable);
			return content.Select(_converter.Convert).ToArray();
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Configuration/AzureAwsOptionsProvider.fs
namespace OrangeJuice.Server.FSharp.Configuration

open System.Threading.Tasks

open Microsoft.WindowsAzure.Storage.Table

open OrangeJuice.Server
open OrangeJuice.Server.Configuration
open OrangeJuice.Server.Services

type AzureAwsOptionsProvider(azureOptions : AzureOptions, azureClient : IAzureClient, converter : IConverter<DynamicTableEntity, AwsOptions>) =
    interface IOptionsProvider<AwsOptions> with
        member this.GetOptions() : Task<AwsOptions[]> =
            let task = async {
                let! options = azureClient.GetEntitiesFromTable<DynamicTableEntity>(azureOptions.AwsOptionsTable) |> Async.AwaitTask
                return options |> Seq.map converter.Convert
                               |> Array.ofSeq
            }
            task |> Async.StartAsTask
